### PR TITLE
Issue #527 Differentiate service configurations

### PR DESCRIPTION
**Pull request by ljacomet**
* Split `ServiceCreationConfiguration` from `ServiceConfiguration` to differentiate between use cases.
* Adapt code to reflect split
* Create new specific type for custom XML parser

### CODE REVIEW

File: 107/src/main/java/org/ehcache/jsr107/ServiceConfigurationParser.java [hunk 1]
```diff
@@ -50,7 +50,7 @@ public URI getNamespace() {
   }
 
   @Override
-  public ServiceConfiguration<Jsr107Service> parse(final Element fragment) {
+  public ServiceCreationConfiguration<Jsr107Service> parse(final Element fragment) {
```
Comment (chrisdennis) [inline 107/src/main/java/org/ehcache/jsr107/ServiceConfigurationParser.java]: Should we rename this type to match it's new role?

File: api/src/main/java/org/ehcache/spi/service/ServiceCreationConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Interface regrouping configuration types to be used when creating a {@link Service}.
+ *
+ * @param <T> the service type this configuration works with
+ *
+ */
+public interface ServiceCreationConfiguration<T extends Service> {
+
+  /**
+   * Indicates which service this configuration works with.
```
Comment (chrisdennis) [inline api/src/main/java/org/ehcache/spi/service/ServiceCreationConfiguration.java]: `Indicates which service type this instance configures`... it's more specific than works with now, right?

File: core/src/main/java/org/ehcache/config/Configuration.java [hunk 3]
```diff
@@ -28,7 +28,7 @@
 
   Map<String, CacheConfiguration<?, ?>> getCacheConfigurations();
 
-  Collection<ServiceConfiguration<?>> getServiceConfigurations();
+  Collection<ServiceCreationConfiguration<?>> getServiceConfigurations();
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/config/Configuration.java]: Should we rename this method to align with it's return type?

File: impl/src/main/java/org/ehcache/internal/store/DefaultStoreProvider.java [hunk 4]
```diff
@@ -26,8 +26,8 @@
 import org.ehcache.internal.store.tiering.CompoundCachingTierServiceConfiguration;
 import org.ehcache.spi.ServiceProvider;
 import org.ehcache.spi.cache.Store;
-import org.ehcache.spi.service.ServiceConfiguration;
 import org.ehcache.spi.service.ServiceDependencies;
+import org.ehcache.spi.service.ServiceConfiguration;
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/internal/store/DefaultStoreProvider.java]: Huh?

File: impl/src/main/java/org/ehcache/internal/store/disk/OffHeapDiskStore.java [hunk 5]
```diff
@@ -38,8 +38,8 @@
 import org.ehcache.spi.serialization.Serializer;
 import org.ehcache.spi.service.FileBasedPersistenceContext;
 import org.ehcache.spi.service.LocalPersistenceService;
-import org.ehcache.spi.service.ServiceConfiguration;
 import org.ehcache.spi.service.ServiceDependencies;
+import org.ehcache.spi.service.ServiceConfiguration;
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/internal/store/disk/OffHeapDiskStore.java]: Huh?

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java [hunk 6]
```diff
@@ -46,8 +46,8 @@
 import org.ehcache.spi.cache.tiering.CachingTier;
 import org.ehcache.spi.serialization.SerializationProvider;
 import org.ehcache.spi.serialization.Serializer;
-import org.ehcache.spi.service.ServiceConfiguration;
 import org.ehcache.spi.service.ServiceDependencies;
+import org.ehcache.spi.service.ServiceConfiguration;
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: Huh?

File: impl/src/main/java/org/ehcache/internal/store/offheap/OffHeapStore.java [hunk 7]
```diff
@@ -33,8 +33,8 @@
 import org.ehcache.spi.cache.tiering.LowerCachingTier;
 import org.ehcache.spi.serialization.SerializationProvider;
 import org.ehcache.spi.serialization.Serializer;
-import org.ehcache.spi.service.ServiceConfiguration;
 import org.ehcache.spi.service.ServiceDependencies;
+import org.ehcache.spi.service.ServiceConfiguration;
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/internal/store/offheap/OffHeapStore.java]: Huh?

File: 107/src/main/java/org/ehcache/jsr107/ServiceConfigurationParser.java (hunk 1 above)
Comment (ljacomet) [inline 107/src/main/java/org/ehcache/jsr107/ServiceConfigurationParser.java]: `JSr107ServiceConfigurationParser` seems better. I will rename though the XML parser interfaces to indicate better what they stand for.

File: api/src/main/java/org/ehcache/spi/service/ServiceCreationConfiguration.java (hunk 2 above)
Comment (ljacomet) [inline api/src/main/java/org/ehcache/spi/service/ServiceCreationConfiguration.java]: `Indicates which service consumes this configuration at creation` ?

**Pull request merged**